### ISSUE TITLE
Can't see database files in SQLite DB Browser

### ISSUE THREAD

**Issue by cheddington** [sqlitebrowser/sqlitebrowser#1648]
I just installed SQLite DB Browser for Mac to view a Safari browsing history database, but when I try to open or browse using DB Browser, the User/[USERNAME]/Library/Safari folder which contains history.db shows a minus sign and does not show any other folders of files within it. Of course, I can see all the folders and files using Finder, but not using SQLite DB Browser. Why can I not see or open the contents of this folder?

Using Mojave 10.14.1.

**Comment by cheddington** [comment]
The entire User/[Username]/Library folder is by default a hidden folder, but I have selected the option to show this folder.

**Comment by chrisjlocke** [comment]
Aah.  Was hoping it was related to #1604, and you could just press <kbd>Command</kbd>+<kbd>Shift</kbd>+<kbd>.</kbd> while in the file dialog...

**Comment by cheddington** [comment]
Pressing Command+Shift+. while in the file dialog doesn't cause the minus sign to disappear from the Safari folder and doesn't allow anything contained in the folder to be seen.

**Comment by cheddington** [comment]
I also tried to navigate to the file in the Finder and selected Open With... and selected DB Browser for SQLite. I received the following message: "Could not open database file. Reason: Invalid file format"

**Comment by chrisjlocke** [comment]
Is Safari closed when you're trying to open the database in DB4S?  Safari will keep its claws in it while open.

**Comment by cheddington** [comment]
It behaves the same with Safari closed as it does when Safari is open. 

I was able to make a copy of the file using the Finder, paste it into a new location and find it in DB Browser for SQLite. I'll remember this workaround for the future. Great recommendation. Thanks.

**Comment by chrisjlocke** [comment]
> It behaves the same with Safari closed as it does when Safari is open.

Dangit.  You could try renaming it, which would confirm the file is closed and readable.  At least you have a workaround, albeit fiddly.

**Comment by justinclift** [comment]
> Safari will keep its claws in it while open.

Guessing here, but suspecting you're thinking of windows file locking approach blocking other processes.  *nix systems do have similar available, but it's not used anywhere near as aggressively as windows does it.

It's fairly unusual on *nix to block other processes from opening files just because they're in use.  Note that it **does** happen where there's good call for it, but it's not a default behaviour. :smile:

> You could try renaming it, which would confirm the file is closed and readable.

Not on *nix (including macOS).  That's a windows-specific dumb behaviour. :wink:

Changing a file name while it's open is very much allowed on *nix, and won't have any negative impact on processes that current have the file open.

Naturally if something else tries to open the file using the old name it'll miss.  But there's no blocking of renames just because someone's using a file.

**Comment by chrisjlocke** [comment]
OK, thanks.  Copying the file worked though, so explain that one! 😉

**Comment by justinclift** [comment]
:man_shrugging:

**Comment by justinclift** [comment]
Hmmm, that more sounds like a macOS specific weirdness where it didn't want to show the contents of the application directory in the dialog for some unknown (for now) reason.  It'll probably turn out to be some setting somewhere (maybe security related?) trying to enforce something.

The workaround of copying the file into a visible-to-DB4S-location seems like a go-er in this instance. :wink:

**Comment by mgrojo** [comment]
@cheddington  I hope we have helped you. I'll close this now, but you can still comment on it if you'd like.

**Comment by joanna350** [comment]
I cannot run the copy command, (with sudo) nor open the manually copied file of the database. Any suggestion..?

**Comment by chrisjlocke** [comment]
Any chance of a screenshot so we can see what you're seeing?

**Comment by joanna350** [comment]
Hi thanks for your reply. Here. I just read somewhere that nowadays the sqlite command does not enable the feature we once expected anymore either, so was about to fold on the matter.. 
<img width="1043" alt="Screenshot 2020-12-10 at 13 18 12" src="https://user-images.githubusercontent.com/19216902/101777298-30c66a00-3aea-11eb-9b00-fae80d54b7ee.png">

**Comment by chrisjlocke** [comment]
Thanks.  Sorry to be a pain, but could you also screenshot *just* before you selected the file, so we can see how you're opening the file?  So just before you got this error message.

> the sqlite command does not enable the feature we once expected anymore either, so was about to fold on the matter

I don't understand this - unsure what feature you're referring to. 😣

**Comment by joanna350** [comment]
Of course, no problem. The path to the directory Safari is luckily the same as mentioned here `User/[USERNAME]/Library/Safari`
<img width="812" alt="Screenshot 2020-12-10 at 21 15 03" src="https://user-images.githubusercontent.com/19216902/101830673-cbe03380-3b2c-11eb-8da3-4ad6154df959.png">

What I mean by feature is that - I thought these timestamp information was either just non-existent or locked for some reason in newer versions.

**Comment by garyking** [comment]
@joanna350 Were you ever able to figure out how to open the `History.db` file in the DB Browser? I have the exact same problem as you, with the exact same file.

**Comment by chrisjlocke** [comment]
Can you copy that history.db (and ancillary files ... history.db-lock, history.db-shm and history.db-wal) to your home directory, does DB4S open it then?

**Comment by garyking** [comment]
> Can you copy that history.db (and ancillary files ... history.db-lock, history.db-shm and history.db-wal) to your home directory, does DB4S open it then?

Yes, I am able to open the database if I copy the database file to my home directory in DB4S.

**Comment by garyking** [comment]
To fix the problem on a Mac to access restricted database files, go to "System Preferences > Security & Privacy > Privacy > Full Disk Access" and add and checkmark DB4S.

**Comment by chrisjlocke** [comment]
Thanks Gary - useful information!

**Comment by cheddington** [comment]
@garyking 

> To fix the problem on a Mac to access restricted database files, go to "System Preferences > Security & Privacy > Privacy > Full Disk Access" and add and checkmark DB4S.
I haven't attempted this since my original posts almost four years ago, but I was able to access and open the Safari history database after applying your fix. However, I reversed the fix (removed the checkmark from the Full Disk Access section of Privacy and was still able to open the database normally. Perhaps this was something Apple did in versions of the operating system since 2018.
